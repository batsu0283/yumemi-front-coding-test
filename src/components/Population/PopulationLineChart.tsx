import {
  CartesianGrid,
  Legend,
  Line,
  LineChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";
import type { PopulationCategoriesValue } from "../../hooks/usePopulationCategories";
import type { Populations } from "../../hooks/usePopulations";
import {
  type PrefCode,
  populationCategories,
  prefectureColors,
} from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";

type PopulationLineChartProps = {
  prefectures: Prefecture[];
  selectedPopulationCategory: PopulationCategoriesValue;
  handlePopulationCategoryChange: (targetValue: string) => void;
  populations: Populations;
  checkedPrefectures: Set<PrefCode>;
};

export const PopulationLineChart = ({
  prefectures,
  selectedPopulationCategory,
  handlePopulationCategoryChange,
  populations,
  checkedPrefectures,
}: PopulationLineChartProps) => (
  <section className="mt-5 mb-10">
    <label className="block text-gray-700 mb-2">
      表示するデータを選択
    <select
      value={selectedPopulationCategory}
      onChange={(e) => handlePopulationCategoryChange(e.target.value)}
      className="block w-full max-w-xs p-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-cyan-500 focus:border-cyan-500"
    >
      {populationCategories.map((populationCategory) => (
        <option key={populationCategory.value} value={populationCategory.value}>
          {populationCategory.label}
        </option>
      ))}
    </select>
    </label>
    <ResponsiveContainer width="100%" height={500}>
      <LineChart
        data={populations[selectedPopulationCategory]}
        margin={{ top: 5, right: 20, left: 30, bottom: 5 }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="year" />
        <YAxis />
        <Tooltip />
        {Array.from(checkedPrefectures).map((prefCode) => (
          <Line
            key={prefCode}
            type="linear"
            dataKey={prefCode}
            stroke={prefectureColors[prefCode]}
            name={
              prefectures.find((pref) => pref.prefCode === prefCode)?.prefName
            }
            unit="人"
          />
        ))}
        <Legend />
      </LineChart>
    </ResponsiveContainer>
  </section>
);
