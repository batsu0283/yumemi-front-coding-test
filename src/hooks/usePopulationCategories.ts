import { useState } from "react";
import { populationCategories } from "../App";

type PopulationCategoriesValue = (typeof populationCategories)[number]["value"];

const isPopulationCategory = (
  value: string,
): value is PopulationCategoriesValue => {
  return populationCategories.some((category) => category.value === value);
};

export const usePopulationCategories = () => {
  const [selectedPopulationCategory, setSelectedPopulationCategory] =
    useState<PopulationCategoriesValue>("total");

  const handlePopulationCategoryChange = (targetValue: string) => {
    if (!isPopulationCategory(targetValue)) return;
    setSelectedPopulationCategory(targetValue);
  };

  return { selectedPopulationCategory, handlePopulationCategoryChange };
};
