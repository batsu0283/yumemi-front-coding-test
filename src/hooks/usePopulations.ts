import { useCallback, useState } from "react";
import type { PrefCode } from "../utils/constants";
import { type YearlyPopulation, resasApi } from "../utils/resasApi";

export type Populations = {
  total: PopulationGraphData[];
  young: PopulationGraphData[];
  workingAge: PopulationGraphData[];
  elderly: PopulationGraphData[];
};

type PopulationGraphData = {
  year: number;
} & {
  [K in PrefCode]?: number;
};

const updatePopulations = (
  prevTotalPopulations: PopulationGraphData[],
  totalData: YearlyPopulation[],
  prefCode: PrefCode,
): PopulationGraphData[] => {
  const newTotalPopulations = [...prevTotalPopulations];
  for (const { year, value } of totalData) {
    const existingEntry = newTotalPopulations.find(
      (entry) => entry.year === year,
    );
    if (existingEntry) {
      existingEntry[prefCode] = value;
    } else {
      newTotalPopulations.push({ year, [prefCode]: value });
    }
  }
  return newTotalPopulations;
};

const updateCheckedPrefectures = (
  prevCheckedPrefectures: Set<PrefCode>,
  prefCode: PrefCode,
) => {
  const newCheckedPrefectures = new Set(prevCheckedPrefectures);
  if (newCheckedPrefectures.has(prefCode)) {
    newCheckedPrefectures.delete(prefCode);
  } else {
    newCheckedPrefectures.add(prefCode);
  }
  return newCheckedPrefectures;
};

export const usePopulations = () => {
  const [checkedPrefectures, setCheckedPrefectures] = useState<Set<PrefCode>>(
    new Set(),
  );
  const [populations, setPopulations] = useState<Populations>({
    total: [],
    young: [],
    workingAge: [],
    elderly: [],
  });

  const handlePrefectureCheckBox = useCallback(
    async (prefCode: PrefCode) => {
      const isAlreadyFetched = populations.total[0]?.[prefCode];
      if (isAlreadyFetched) {
        setCheckedPrefectures((prevCheckedPrefectures) =>
          updateCheckedPrefectures(prevCheckedPrefectures, prefCode),
        );
        return;
      }
      const queryParams = new URLSearchParams({
        prefCode: prefCode.toString(),
        cityCode: "-", // 全市区町村
      });
      try {
        const data = await resasApi.getPopulationCompositionPerYear(
          queryParams.toString(),
        );
        setPopulations((prevPopulations) => ({
          total: updatePopulations(
            prevPopulations.total,
            data.result.data[0].data,
            prefCode,
          ),
          young: updatePopulations(
            prevPopulations.young,
            data.result.data[1].data,
            prefCode,
          ),
          workingAge: updatePopulations(
            prevPopulations.workingAge,
            data.result.data[2].data,
            prefCode,
          ),
          elderly: updatePopulations(
            prevPopulations.elderly,
            data.result.data[3].data,
            prefCode,
          ),
        }));
      } catch {
        window.alert("データの取得に失敗しました");
        return;
      }

      setCheckedPrefectures((prevCheckedPrefectures) =>
        updateCheckedPrefectures(prevCheckedPrefectures, prefCode),
      );
    },
    [populations.total[0]],
  );

  return { checkedPrefectures, populations, handlePrefectureCheckBox };
};
