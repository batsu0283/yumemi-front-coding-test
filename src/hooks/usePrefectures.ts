import { useEffect, useState } from "react";
import { type Prefecture, resasApi } from "../utils/resasApi";

export const usePrefectures = () => {
  const [prefectures, setPrefectures] = useState<Prefecture[]>([]);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const fetchPrefectures = async () => {
      try {
        const data = await resasApi.getPrefectures();
        setPrefectures(data.result);
      } catch {
        setIsError(true);
      }
    };
    fetchPrefectures();
  }, []);

  return { prefectures, isError };
};
