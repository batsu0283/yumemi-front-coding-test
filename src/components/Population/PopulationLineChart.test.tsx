import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";
import { usePopulationCategories } from "../../hooks/usePopulationCategories";
import type { Populations } from "../../hooks/usePopulations";
import type { PrefCode } from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";
import { PopulationChart } from "./PopulationLineChart";

vi.mock("../../hooks/usePopulationCategories");

describe("PrefecturePopulationChart", () => {
  const mockPrefectures: Prefecture[] = [
    { prefCode: 1, prefName: "北海道" },
    { prefCode: 2, prefName: "青森県" },
  ];

  const mockPopulations: Populations = {
    total: [{ year: 2020, 1: 1000, 2: 1500 }],
    young: [{ year: 2020, 1: 200, 2: 300 }],
    workingAge: [{ year: 2020, 1: 600, 2: 900 }],
    elderly: [{ year: 2020, 1: 200, 2: 300 }],
  };

  const mockCheckedPrefectures = new Set<PrefCode>([1, 2]);

  const mockedUsePopulationCategories = vi.mocked(usePopulationCategories);
  afterEach(() => {
    vi.restoreAllMocks();
  });

  beforeAll(() => {
    globalThis.ResizeObserver = class ResizeObserver {
      observe() {
        console.log("observe");
      }
      unobserve() {
        console.log("unobserve");
      }
      disconnect() {
        console.log("disconnect");
      }
    };
  });

  it("PopulationCategorySelector と PopulationLineChartContent が表示されること", () => {
    mockedUsePopulationCategories.mockReturnValue({
      selectedPopulationCategory: "total",
      handlePopulationCategoryChange: vi.fn(),
    });
    const { container } = render(
      <PopulationChart
        prefectures={mockPrefectures}
        populations={mockPopulations}
        checkedPrefectures={mockCheckedPrefectures}
      />,
    );
    expect(screen.getByText("表示するデータを選択")).toBeInTheDocument();
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    const rechartsContainer = container.querySelector(
      ".recharts-responsive-container",
    );
    expect(rechartsContainer).toBeInTheDocument();
  });

  it("カテゴリ変更時に handlePopulationCategoryChange が呼ばれること", () => {
    const mockHandlePopulationCategoryChange = vi.fn();
    mockedUsePopulationCategories.mockReturnValue({
      selectedPopulationCategory: "total",
      handlePopulationCategoryChange: mockHandlePopulationCategoryChange,
    });
    render(
      <PopulationChart
        prefectures={mockPrefectures}
        populations={mockPopulations}
        checkedPrefectures={mockCheckedPrefectures}
      />,
    );

    const select = screen.getByRole("combobox");
    fireEvent.change(select, { target: { value: "young" } });

    expect(mockHandlePopulationCategoryChange).toHaveBeenCalledWith("young");
    expect(mockHandlePopulationCategoryChange).toHaveBeenCalledTimes(1);
  });

  it("Snapshot", () => {
    const mockHandlePopulationCategoryChange = vi.fn();
    mockedUsePopulationCategories.mockReturnValue({
      selectedPopulationCategory: "total",
      handlePopulationCategoryChange: mockHandlePopulationCategoryChange,
    });
    const { container } = render(
      <PopulationChart
        prefectures={mockPrefectures}
        populations={mockPopulations}
        checkedPrefectures={mockCheckedPrefectures}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
