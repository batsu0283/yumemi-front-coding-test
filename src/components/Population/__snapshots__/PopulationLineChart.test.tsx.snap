// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefecturePopulationChart > Snapshot 1`] = `
<div>
  <section
    class="mt-5 mb-10"
  >
    <label
      class="block text-gray-700 mb-2"
    >
      表示するデータを選択
      <select
        class="block w-full max-w-xs p-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-cyan-500 focus:border-cyan-500"
      >
        <option
          value="total"
        >
          総人口
        </option>
        <option
          value="young"
        >
          年少人口
        </option>
        <option
          value="workingAge"
        >
          生産年齢人口
        </option>
        <option
          value="elderly"
        >
          老年人口
        </option>
      </select>
    </label>
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 500px; min-width: 0;"
    />
  </section>
</div>
`;
